keras_nlp/utils/text_generation.py

```diff
@@ -65,6 +71,90 @@ def mask_tokens_after_end_token(prompt, max_length, end_token_id, pad_token_id):
     return tf.where(valid_indices, prompt, pad_token_id)
 
 
+def _generate_text(
```

Comment:
Overall this feels a little hard to follow. I wonder if we can unify the flow between ragged and dense, and simplify the state. Something like...

```
...outside the loop...
# Make sure the prompt is ragged.
if isinstance(prompt, tf.Tensor):
   prompt = tf.RaggedTensor.from_tensor(prompt)
# Convert the prompt to dense with shape (bs, max_length), and keep a mask.
mask = tf.ones_like(prompt, dtype=tf.bool).to_tensor(...)
prompt = prompt.to_tensor(...)

...inside the loop...
next_column = gen_next_token_fn(prompt[:, :length])
# Make sure to not replace the original prompt input.
next_column = tf.where(mask[:, length + 1], prompt[:, length + 1], next_column)
prompt = tf.tensor_scatter_nd_update(...)
```

Not sure if that's totally clear above, but basically the idea is that
1) The ragged case is more general, so we can write one branch of code by converting all prompts to raggeds.
2) We should also be able to simplify the code and the `token_probability_fn` by always calling the `token_probability_fn` on a dense tensor and masking out certain updates.

Reply:
Have not condensed it to one branch because `.to_tensor()` does not work with XLA. Made changes for the masking logic.